### ISSUE TITLE
Timestamp management of RESTART Option

### ISSUE THREAD

**Issue by AaronDonahue** [E3SM-Project/scream#652]
We need a way in the atmosphere driver to determine if this run is a restart or initial simulation.

Following the breadcrumbs in how EAM (cam) determines this information it looks like it is gathered from the component coupler, via the `start_type` string argument.   See `components/cam/src/cpl/atm_comp_mct.F90` line 257

```
! 
       ! Get data from infodata object
       !
       call seq_infodata_GetData( infodata,                                           &
            case_name=caseid, case_desc=ctitle,                                       &
            start_type=starttype, 
...
```

start_type is a string that can have 1 of 3 options that are defined in the `cime/src/drivers/mct/shr/seq_infodata_mod.F90` module.  These correspond to "startup"->initial, "continue"->restart, "branch"->branched run (kind of like a specialized restart).

The EAM-ithic way to process this info is to set the time manager according to the flag.  In the case of initial then initialize the time manager.  In the other two cases, the timestamp info is gathered from the component coupler and the timestamp is set.

Additionally, the time manager keeps track of the number of global model steps, which is 0 for an initial run and >0 for restart or branch.  This can be used to synchronize restart files and to trigger any restart type initialization.

SCREAM will need to have some equivalent way to gather timestamp info from the component coupler.  An additional thread of complexity, is that for unit testing we will also need a way to set the timestamp info without access to the component coupler.

**Comment by AaronDonahue** [comment]
@jeff-cohere and @bartgol , I assigned you both because of your knowledge of the timestamp utility.

**Comment by bartgol** [comment]
Can you clarify where/why the atm needs to know whether it is a restart or startup simulation? I'm assuming the only difference is that restarts need to read some stuff in from file.

Is there any other reason? Do you need/want to save output not with timestamps 0,1,.. but n,n+1,.. maybe?

**Comment by bartgol** [comment]
If those are the only reasons, then no change is really needed to the TimeStamp class. Instead, the AD simply has to get that info (restart and n0) from the atm_mct_init subroutine, and pass them to the I/O classes.

**Comment by AaronDonahue** [comment]
The needs that come immediately to mind are:
IO:  Which needs to know if a restart run is picking up in the middle of an averaging range, at which point it needs to read in a restart history file.  Also, output needs to know which integer timestep we are in to determine if it is appropriate to write output at the frequency requested, if the choice was every nth step.
AD: Will need to know if it needs to read in a restart file to populate the appropriate fields.
TimeStamp: Keeps track of the time of simulation, it will need to start off at the appropriate timestamp which won't be "0" for a restart run.  For example, if RAD queries the TimeStamp to determine the season and angle of the sun it has to be accurate.
Processes themselves:  We currently allow procs to initialize fields at the first step, but if it is a restart run then they shouldn't, instead the AD should take care of that, and if a field that should have been initialized isn't than we should throw an error.

There may be others.

**Comment by bartgol** [comment]
Quick note: the time stamp already requires a valid yymmddss initialization. There's no default TimeStamp constructor (and definitely not at 0A.D.). The `atm_mct_init` routine is responsible to fwd that to the AD.


Also, the atm procs init their fields as requested in the input file. Do you want to overrule input file specs for restart runs (or even error out)? That is, if it's a restart run, but the user distractingly sets options in the input file for initing inputs; what should the AD do?

**Comment by AaronDonahue** [comment]
Ok good, so the infrastructure to ensure that the timestamp matches the component coupler is already in place.  Per your second question, I think throwing an error would be best if a user put options in the input file for a restart run.  I support more communication through error messages than having the AD overrule the input file and have the user "think" that the fields are initialized as specified.

If, in the future, we find a use case where a user will need to init a field even during a restart we can make an adjustment, but I see this as a unique use case.

**Comment by jeff-cohere** [comment]
So what's to be done here? Checking input for overridden values in the case of restarts, and reporting errors if such overridden values are found? Anything else?

**Comment by bartgol** [comment]
I'm ok with throwing if user asks, say, P3 to init inputs in a restart run. It's definitely safe, and until we have a reason to allow this scenario, let's explicitly prohibit it.

I'm ok putting the timestep number in TimeStamp, so that the `t0` passed to the `initialize` method of atm proc can help spotting the error. It is not the "cleanest" solution (a timestamp is universal, regardless of time discretization specs; a timestep number has to do with timestepping), but it's by far the easiest.

**Comment by AaronDonahue** [comment]
Jeff, if we have someway to query if we are a restart run.  For example a getter function that returns a bool then all we have to do is throw an error if say `restart=true` and the user has asked to initialize any variables.  

As for the timestep, definitely this can be an abstract concept from the perspective of sub-stepping.  But I'm pretty sure there is a universal time step number associated with the number of times the ATM has been stepped by the component coupler.  It is probably safest to pin our timestep number to that value and update it by +1 whenever the surface coupling is called.  Of course when it comes to unit tests with no surface coupler we will want to have a different strategy.

**Comment by PeterCaldwell** [comment]
I think we should close this issue in favor of #798, which I think is a more modern version of the same thing.

**Comment by AaronDonahue** [comment]
I agree that this issue could just be merged into #798
